graphbolt/src/cuda/extension/unique_and_compact_map.cu

```diff
@@ -0,0 +1,256 @@
+/**
+ *  Copyright (c) 2023, GT-TDAlab (Muhammed Fatih Balin & Umit V. Catalyurek)
+ * @file cuda/unique_and_compact_map.cu
+ * @brief Unique and compact operator implementation on CUDA using hash table.
+ */
+#include <graphbolt/cuda_ops.h>
+#include <thrust/gather.h>
+
+#include <cuco/static_map.cuh>
+#include <cuda/std/atomic>
+#include <numeric>
+
+#include "../common.h"
+#include "../utils.h"
+#include "./unique_and_compact.h"
+
+namespace graphbolt {
+namespace ops {
+
+constexpr int node_id_bits = 40;
+
+template <typename index_t, typename map_t>
+__global__ void _InsertAndSetMinBatched(
+    const int64_t num_edges, const int32_t* const indexes, index_t** pointers,
+    const int64_t* const offsets, map_t map) {
+  int64_t i = blockIdx.x * blockDim.x + threadIdx.x;
+  const int stride = gridDim.x * blockDim.x;
+
+  while (i < num_edges) {
+    const int64_t tensor_index = indexes[i];
+    const auto tensor_offset = i - offsets[tensor_index];
+    const int64_t node_id = pointers[tensor_index][tensor_offset];
+    const auto batch_index = tensor_index / 2;
+    const int64_t key = node_id | (batch_index << node_id_bits);
```

Comment:
Will it overflow when `node_ids_bits` is 64?
It's ok we don't handle case node ids > 2**32, but it need to be noted in the code.

Reply:
We support graphs with less than trillion nodes.

Reply:
Meaning node_ids > 2**40 are not supported. I know graphs with trillion edges but I don't know any graphs with trillion nodes.